text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,1131 @@
+no unintended modifications are present and validate that each `Remove` action on main resources has a corresponding
+`Add` of the same resource type. Additionally validate that the `beforeContext` and `afterContext` values for each
+`Remove` and `Add` pair are identical, as CloudFormation stack refactors prohibit configuration changes.
+
+For the `Retain-Remove-Import Migration` strategy, we will introduce an additional option `--import` in the
+`cdk diff` command like `cdk diff --json --import` which will create a change set on the existing stack with
+`import` change set type. An example is as follows:
+
+```json
+$ cdk diff --json --import
+[
+  {
+    "type": "Resource",
+    "resourceChange": {
+      "action": "Import", // NOTE THAT THIS SHOWS "IMPORT"
+      "logicalResourceId": "MyTable794EDED1",
+      "physicalResourceId": "DemoStack-MyTable794EDED1-11W4MR8VZ0UPE",
+      "resourceType": "AWS::DynamoDB::GlobalTable",
+      "replacement": "True",
+      "scope": [],
+      "details": [],
+      "afterContext": "..."
+    }
+  },
+  {
+    "type": "Resource",
+    "resourceChange": {
+      "policyAction": "Delete",
+      "action": "Remove",
+      "logicalResourceId": "MyTable794EDED1",
+      "physicalResourceId": "DemoStack-MyTable794EDED1-11W4MR8VZ0UPE",
+      "resourceType": "AWS::DynamoDB::Table",
+      "scope": [],
+      "details": [],
+      "beforeContext": "..."
+    }
+  }
+]
+```
+
+The `construct-upgrade` CLI command will implicitly call `cdk diff --json --import` which will create an `import`
+type change set as above and it will validate if the main resources' action is `Import` instead of `Add`.
+
+#### [In-Place Migration Only] CloudFormation stack refactoring Validation
+
+For in-place migrations, the construct-upgrade CLI command ensures the correctness of logical ID mappings
+defined in `refactor.json` by validating against the deployed V1 template and synthesized V2 template.
+
+```json
+[
+    {
+        "Source": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcv1A85508D7"
+        },
+        "Destination": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcA2121C38"
+        }
+    },
+    {
+        "Source": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcv1ipv6cidr87175A16"
+        },
+        "Destination": {
+            "StackName": "<stack-name>",
+            "LogicalResourceId": "vpcamazonProvided373FDA5C"
+        }
+    },
+    ......
+]
+```
+
+The validation begins by verifying that every source LogicalId in `refactor.json` corresponds to a resource
+in the deployed V1 template, while every target LogicalId maps to a resource in the synthesized V2 template.
+
+Orphaned entries — such as unmapped logical IDs or resources missing from either template — will block the upgrade.
+Resource types must also align: for example, a resource in V1 template of type `AWS::EC2::VPC` can only map to
```

Comment:
How do we determine which types are compatible to which types? is this construct specific logic?

Reply:
I think this is a general context info that we have (not specific to this validation step). For supported construct upgrade, i.e. VPC or Table, we should have a generic context information on the relevant main resources and we have to map main resource type to the same main resource type in two templates.